### ISSUE TITLE
Stripe 404 when submitting a test payment

### ISSUE THREAD

**Issue by RoryEnright** [ga-wdi-boston/group-project#354]
Using the tutorial from [Stripe Docs](https://stripe.com/docs/recipes/custom-checkout) When I submit the payment info, I'm getting a 404 not found as the .catch response.  

Backend server error:
```
Error: Cannot find module 'app/middleware'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/rory/wdi/group-project/e-commerce-back-end/server.js:8:20)
```
This makes me unsure is many of the required paths are correct, but I'm hesitant to change the paths just for Stripe in fear of goofing up the entire back end.  I tried fixing the paths on my own branch and it eventually fails at:
```
(node:23853) DeprecationWarning: `open()` is deprecated in mongoose >= 4.11.0, use `openUri()` instead, or set the `useMongoClient` option if using `connect()` or `createConnection()`
```
This is an error with the latest mongoose version?  I feel like this is probably just a rabbit hole and there is a different error at play here.

Front End
``` 
<h2>Stripe Checkout Example</h2>

    <div id="shop">
      <button id="buttonCheckout">Checkout</button>
    </div>
<script>
// configure checkout
const checkoutHandler = StripeCheckout.configure({
  key: 'pk_test_6sTEST KEYy',
  locale: 'auto'
})

// add event listener to the checkout button
const button = document.getElementById('buttonCheckout')
button.addEventListener('click', function (ev) {
  checkoutHandler.open({
    name: 'Sample Store',
    description: 'Example Purchase',
    token: handleToken
  })
})

// Token is a callback that Checkout triggers.  Create JSON with token for
// an ajax request

function handleToken (token) {
  $.ajax('/charge', {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify(token)
  })
  .then(output => {
    if (output.status === 'succeeded')
      document.getElementById('shop').innerHTML = '<p>Purhcase complete!</p>'
  })
  .catch(err => {
    console.log('Purchase failed:', err)
  })
}
</script>
```

Back End
```
const express = require('express')
const app = express()
const middleware = require('./app/middleware')
// -------------------------------------------------------------------
// Stripe Keys
const keyPublishable = process.env.PUBLISHABLE_KEY
const keySecret = process.env.SECRET_KEY
// Stripe module with secret key
const stripe = require('stripe')(keySecret)
const bodyParser = require('body-parser')

app.use(express.static('public'))

app.use(bodyParser.urlencoded({extended: false}))
app.use(bodyParser.json())

app.post('/charge', (req, res) => {
  amount = 500

  stripe.customers.create({
    email: req.body.email,
    card: req.body.id
  })
  .then(customer =>
    stripe.charges.create({
      amount,
      description: 'Sample Charge',
      currency: 'usd',
      customer: customer.id
    }))
  .then(charge => res.send(charge))
  .catch(err => {
    console.log('Error:', err)
    res.status(500).send({error: 'Purchase Failed'})
  })
})

app.listen(8000)
```

**Comment by RoryEnright** [comment]
```
POST http://localhost:7165/charge 404 (Not Found)
```
Should this be posting to the stripe api?  Are we supposed to have our own table for charges?

**Comment by RoryEnright** [comment]
UPDATE:  switched around the back end.
Controller:
```
 const stripe = require('stripe')(process.env.STRIPE_SECRET_KEY)
const amount = 500

const charge = (req, res) => {
  stripe.customers.create({
    email: req.body.email,
    card: req.body.id
  })
  .then(customer =>
    stripe.charges.create({
      amount,
      description: 'Sample Charge',
      currency: 'usd',
      customer: customer.id
    }))
  .then(charge => res.send(charge))
  .catch(err => {
    console.log('Error:', err)
    res.status(500).send({error: 'Purchase Failed'})
  })
}
```
Routes:
.post('/charge', 'users#charge')

**Comment by RoryEnright** [comment]
I've gone through previous cohort projects, searched google 'for stripe routes 404', 'stripe express set up', been to stackOverflow threads, watched youtube tutorials and have tried reading the stripe docs.  I'm either missing one small piece or have been chasing the wrong answer for the past two days.

**Comment by jordanallain** [comment]
whats the current state of this issue @RoryEnright ? is it still the same error as earlier?

**Comment by tvcrawley** [comment]
@RoryEnright Is this still an issue?

**Comment by laurpaik-zz** [comment]
@RoryEnright @slammyde7113 @isharma-sharma

Does this help? https://stripe.com/docs/checkout#integration-custom

**Comment by slammyde7113** [comment]
@laurpaik we looked at the custom input page and tried using that set-up and now we're sending some data. We're getting a 402 request from the server now.
```
<script src="https://checkout.stripe.com/checkout.js"></script>

<button id="customButton">Purchase</button>

<script>
var handler = StripeCheckout.configure({
  key: 'pk_test_*********************',
  image: 'https://stripe.com/img/documentation/checkout/marketplace.png',
  locale: 'auto',
  token: function(token) {
    // You can access the token ID with `token.id`.
    // Get the token ID to your server-side code for use.
  }
});

document.getElementById('customButton').addEventListener('click', function(e) {
  // Open Checkout with further options:
  handler.open({
    name: 'Nozama.com',
    description: '2 widgets',
    amount: 2000
  });
  e.preventDefault();
});

// Close Checkout on page navigation:
window.addEventListener('popstate', function() {
  handler.close();
});
</script>
```

**Comment by slammyde7113** [comment]
Progress!!

**Comment by RoryEnright** [comment]
402 was due to a bad credit card test date.  Make sure you don't go too far out, was giving it a year of 2022.  Keep the year within 4 years of current year.  

Did a console.log(token.id) and saw that we were getting the appropriate token from the response.
Set token = token.id

  token: function(token) {
    const token = token.id
  }

Next step is to use that token to make a charge.

curl scripts with the token we got back works!
curl https://api.stripe.com/v1/charges \
   -u sk_test_your key here \
   -d amount=2000 \
   -d currency=usd \
   -d source= *THE TOKEN ID FROM RESPONSE GOES HERE* \
